src/engine/SpatialJoinAlgorithms.cpp

```diff
@@ -223,3 +224,259 @@ Result SpatialJoinAlgorithms::S2geometryAlgorithm() {
+      boxes.emplace_back(point(-180, -90), point(rightBound, lowerBound));
+    } else {
+      boxes.emplace_back(point(leftBound, -90), point(rightBound, lowerBound));
+    }
+  }
+  // add the box(es) inbetween the longitude lines
+  if (boxCrosses180Longitude) {
+    // only one box needed to cover the longitudes
+    boxes.emplace_back(point(rightBound, -90), point(leftBound, 90));
+  } else {
+    // two boxes needed, one left and one right of the anti bounding box
+    boxes.emplace_back(point(-180, -90), point(leftBound, 90));
+    boxes.emplace_back(point(rightBound, -90), point(180, 90));
+  }
+  return boxes;
+}
+
+// ____________________________________________________________________________
+bool SpatialJoinAlgorithms::containedInBoundingBoxes(
+    const std::vector<box>& bbox, point point1) const {
+  // correct lon bounds if necessary
+  while (point1.get<0>() < -180) {
+    point1.set<0>(point1.get<0>() + 360);
+  }
+  while (point1.get<0>() > 180) {
+    point1.set<0>(point1.get<0>() - 360);
+  }
+  if (point1.get<1>() < -90) {
+    point1.set<1>(-90);
+  } else if (point1.get<1>() > 90) {
+    point1.set<1>(90);
+  }
+
+  if (std::ranges::any_of(bbox, [point1](const box& aBox) {
+        return boost::geometry::covered_by(point1, aBox);
+      })) {
+    return true;
+  }
+  return false;
+}
+
+// ____________________________________________________________________________
+Result SpatialJoinAlgorithms::BoundingBoxAlgorithm() {
+  const auto [idTableLeft, resultLeft, idTableRight, resultRight, leftJoinCol,
+              rightJoinCol, numColumns, maxDist, maxResults] = params_;
+  IdTable result{numColumns, qec_->getAllocator()};
+
+  // create r-tree for smaller result table
+  auto smallerResult = idTableLeft;
+  auto otherResult = idTableRight;
+  bool leftResSmaller = true;
+  auto smallerResJoinCol = leftJoinCol;
+  auto otherResJoinCol = rightJoinCol;
+  if (idTableLeft->numRows() > idTableRight->numRows()) {
+    smallerResult = idTableRight;
+    otherResult = idTableLeft;
+    leftResSmaller = false;
+    smallerResJoinCol = rightJoinCol;
+    otherResJoinCol = leftJoinCol;
+  }
+
+  bgi::rtree<value, bgi::quadratic<16>> rtree;
+  for (size_t i = 0; i < smallerResult->numRows(); i++) {
+    // get point of row i
+    // ColumnIndex smallerJoinCol = getJoinCol(smallerChild, smallerVariable);
+    auto geopoint = getPoint(smallerResult, i, smallerResJoinCol);
+    point p(geopoint.value().getLng(), geopoint.value().getLat());
+
+    // add every point together with the row number into the rtree
+    rtree.insert(std::make_pair(p, i));
+  }
+  for (size_t i = 0; i < otherResult->numRows(); i++) {
+    auto geopoint1 = getPoint(otherResult, i, otherResJoinCol);
+    point p(geopoint1.value().getLng(), geopoint1.value().getLat());
+
+    // query the other rtree for every point using the following bounding box
+    std::vector<box> bbox = computeBoundingBox(p);
+    std::vector<value> results;
```

Comment:
1. Can you use a vector of pointers or something of the like as the type of the `results` vector to save memory?

2. We should make the large data structures respect the memory limit (all classes that take an allocator can take `ad_utility::AllocatorWithMemoryLimit`, see how other operations do this.

Reply:
1. No that's not possible: the type of the vector has to be the same type as the rtree. Therefore the only option would be to change the type of the rtree to a pointer. That's not allowed though.
2. For some reason i don't quite understand the allocator isn't working for me. I think it has something to do with the types of the templates being boost specific types, but i'm not sure about that. Perhaps you can take a look at it